apps/src/lab2/views/ExtraLinks.tsx

```diff
@@ -14,61 +16,75 @@ interface PermissionResponse {
   permissions: string[];
 }
 
-interface ExtraLinksResponse {
-  links: {[key: string]: {text: string; url: string; access_key?: string}[]};
-  can_clone: boolean;
-  can_delete: boolean;
-  level_name: string;
-  script_level_path_links: {
-    script: string;
-    path: string;
-  }[];
-}
-
 // If the user has permission to see extra links, fetch extra links for the level,
 // then display a modal with the link data.
 const ExtraLinks: React.FunctionComponent<ExtraLinksProps> = ({
   levelId,
 }: ExtraLinksProps) => {
   const {loading, data} = useFetch('/api/v1/users/current/permissions');
   const [isModalOpen, setIsModalOpen] = useState(false);
-  const [linkData, setLinkData] = useState<ExtraLinksResponse | null>(null);
+  const [levelLinkData, setLevelLinkData] =
+    useState<ExtraLinksLevelData | null>(null);
+  const [projectLinkData, setProjectLinkData] =
+    useState<ExtraLinksProjectData | null>(null);
+  const permissionData = data
+    ? (data as PermissionResponse)
+    : {permissions: []};
+  const channelId = useAppSelector(
+    state => state.lab.channel && state.lab.channel.id
+  );
 
   useEffect(() => {
```

Comment:
Looking at this again, might it be cleaner to pull out all this fetching logic into a separate function outside of the component, since all three of these server calls have to happen back to back? For example:

```
async function fetchExtraLinksData(levelId: number): Promise<ExtraLinksWithPermissions> {
  // Fetch permissions
  const permissionsResponse = await HttpClient.fetchJson<PermissionResponse>(
    '/api/v1/users/current/permissions'
  );
  const {permissions} = permissionsResponse.value;

  // Fetch level links data
  let levelLinkData: ExtraLinksLevelData;
  if (permissions.includes(PERMISSIONS.LEVELBUILDER)) {
    const levelLinkDataResponse =
      await HttpClient.fetchJson<ExtraLinksLevelData>(
        `/levels/${levelId}/extra_links`
      );
    levelLinkData = levelLinkDataResponse.value;
  }

  // Fetch project links data
  let levelProjectData: ExtraLinksProjectData;
  if (permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)) {
    const levelProjectDataResponse =
      await HttpClient.fetchJson<ExtraLinksProjectData>(
        `/projects/${levelId}/extra_links`
      );
    levelProjectData = levelProjectDataResponse.value;
  }

  // Return fetched data and permissions
  return {
    levelLinkData,
    levelProjectData,
    permissions,
  };
}
```

And then, in the component we just need something like

```
  const [isLoading, setIsLoading] = useState(false);
  const [extraLinksWithPermissions, setExtraLinksWithPermissions] = useState(null);

  useEffect(() => {
    setIsLoading(true);
    fetchExtraLinksData(levelId).then(data => {
      setExtraLinksWithPermissions(data);
      setIsLoading(false);
    });
  }, [levelId]);

  // Do stuff with extraLinksWithPermissions once it exists
```

(couldn't think of a good name for the "extraLinksWithPermissions" bundle, but I'm sure there's something better)

Reply:
Nice! Will update.